client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -444,44 +335,133 @@ private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
+     * this will return the dead host that matches the {@link HostSelector}
+     * that is closest to being revived.
+     * <p>
+     * If no living and no dead hosts match the selector we retry a few
+     * times to handle concurrent modifications of the list of dead hosts.
+     * We never block the thread or {@link Thread#sleep} or anything like
+     * that. If the retries fail this throws a {@link IOException}.
+     * @throws IOException if no hosts are available
      */
-    private HostTuple<Iterator<HttpHost>> nextHost() {
-        final HostTuple<Set<HttpHost>> hostTuple = this.hostTuple;
-        Collection<HttpHost> nextHosts = Collections.emptySet();
+    private HostTuple<Iterator<HttpHost>> nextHost(HostSelector hostSelector) throws IOException {
+        int attempts = 0;
+        NextHostsResult result;
+        /*
+         * Try to fetch the hosts to which we can send the request. It is possible that
+         * this returns an empty collection because of concurrent modification to the
+         * blacklist.
+         */
         do {
-            Set<HttpHost> filteredHosts = new HashSet<>(hostTuple.hosts);
-            for (Map.Entry<HttpHost, DeadHostState> entry : blacklist.entrySet()) {
-                if (System.nanoTime() - entry.getValue().getDeadUntilNanos() < 0) {
-                    filteredHosts.remove(entry.getKey());
+            final HostTuple<Set<HttpHost>> hostTuple = this.hostTuple;
+            result = nextHostsOneTime(hostTuple, blacklist, lastHostIndex, System.nanoTime(), hostSelector);
+            if (result.hosts == null) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("No hosts avialable. Will retry. Failure is " + result.describeFailure());
                 }
+            } else {
+                // Success!
+                return new HostTuple<>(result.hosts.iterator(), hostTuple.authCache, hostTuple.metaResolver);
             }
-            if (filteredHosts.isEmpty()) {
-                //last resort: if there are no good host to use, return a single dead one, the one that's closest to being retried
-                List<Map.Entry<HttpHost, DeadHostState>> sortedHosts = new ArrayList<>(blacklist.entrySet());
-                if (sortedHosts.size() > 0) {
-                    Collections.sort(sortedHosts, new Comparator<Map.Entry<HttpHost, DeadHostState>>() {
-                        @Override
-                        public int compare(Map.Entry<HttpHost, DeadHostState> o1, Map.Entry<HttpHost, DeadHostState> o2) {
-                            return Long.compare(o1.getValue().getDeadUntilNanos(), o2.getValue().getDeadUntilNanos());
-                        }
-                    });
-                    HttpHost deadHost = sortedHosts.get(0).getKey();
+            attempts++;
+        } while (attempts < MAX_NEXT_HOSTS_ATTEMPTS);
+        throw new IOException("No hosts available for request. Last failure was " + result.describeFailure());
```

Comment:
Yeah, it doesn't give any context because sometimes we're reviving a dead node and sometimes we're picking from a list of living ones. 

It *feels* right to me for `HostSelector`s to make very simple choices so it is easier to reason about what they do.

I understand wanting to make the "what if we reject all the hosts" behavior pluggable, but I think it is pretty complicated because you don't know if you rejected all of the hosts or just all of the living hosts.

What if we make our own subclass of `IOException` and folks can catch it and retry without the selector if they really want to? That feels like an exceptional case so I don't feel too bad about it. I'm not sure though.

Reply:
I am not sure that it makes a difference whether we are selecting among all live nodes or trying to revive from the blacklist. If the selector is acting as a preference (like  a rack aware selector), it would still select whatever node if it doesn't find what it likes. But it would know more if we provided all the possible nodes. Let's say we throw exception ourselves whenever the selector returns empty, we document that the selector should never throw an exception, but the selector has more freedom to do what it likes this way. I don't think this complicates things more than what we do now.

Reply:
actually I think would simplify things as we don't even have to wonder whether we need to retry or throw exception. If the selector doesn't return any host we just throw, it is responsibility of the selector to return at least one host given one or more hosts. Knowing that,  a selector can fall back to some backup plan in case e.g. it doesn't find any non master only node.

Reply:
we could even stop getting a single node from the blacklist to be resurrected and just provide the whole blacklist to the selector in case there are no alive hosts (order would still matter).  But if the selector returns empty, I wouldn't try to resurrect, just throw and document well this behaviour.